### PR TITLE
Add Flask Scope Manager

### PR DESCRIPTION
**Pull request by rmfitzpatrick**
Adds a FlaskScopeManager using http://flask.pocoo.org/docs/1.0/api/?highlight=request_ctx_stack#flask._request_ctx_stack as currently done by https://github.com/opentracing-contrib/python-flask.

### CODE REVIEW

**Comment by carlosalberto** [comment]
Hey @rmfitzpatrick 

Thanks for the PR and sorry for the delayed reply ;) I had been playing with the Flask instrumentation a few weeks ago, so I'm definitely curious. I'm wondering if a 'pure' thread-local based solution would work for Flask?

Is there any reason to really need to use always their `stack`? Let me know ;)

**Comment by rmfitzpatrick** [comment]
@carlosalberto, while I think in many use cases a ThreadLocalScopeManager would work, http://werkzeug.pocoo.org/docs/0.14/local/ provides a cursory description of how this is likely not a best practice.  Since Werkzeug supports context locals in different worker types/server configurations relying on its internals seems safer.

File: opentracing/scope_managers/flask.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+class FlaskScopeManager(ScopeManager):
+    """
+    :class:`~opentracing.ScopeManager` implementation for **Flask**
+    that stores the :class:`~opentracing.Scope` in a Flask
+    :class:`RequestContext`, made accessible via
+    :attr:`flask._request_ctx_stack`.
+    """
+    def __init__(self):
+        self._tls = threading.local()
```
Comment (carlosalberto) [inline opentracing/scope_managers/flask.py]: Any reason not to inherit directly from `ThreadLocalScopeManager` and fallback to the base implementation when there's no `stack.top`?

**Comment by carlosalberto** [comment]
Hey @rmfitzpatrick thanks for the PR - and so sorry for the delayed review.

Overall it looks good to me, and I'd discard the examples under `testbed`, as they are basically following the thread-local usage (with the optimization of using `stack.top` directly from Flask).

@yurishkuro Any comment on this? ;)

File: opentracing/scope_managers/flask.py (hunk 1 above)
Comment (rmfitzpatrick) [reply opentracing/scope_managers/flask.py]: Are you suggesting this?
```
-class FlaskScopeManager(ScopeManager):
+class FlaskScopeManager(ThreadLocalScopeManager):
     """
     :class:`~opentracing.ScopeManager` implementation for **Flask**
     that stores the :class:`~opentracing.Scope` in a Flask
     :class:`RequestContext`, made accessible via
     :attr:`flask._request_ctx_stack`.
     """
-    def __init__(self):
-        self._tls = threading.local()
-
     @property
     def _context(self):
-        # Default to threading.local for usage outside
+        # Default to ThreadLocalScopeManager._tls_scope for usage outside
         # app/request context (unit tests)
         if stack.top is None:
-            return self._tls
+            return self._tls_scope
         return stack.top
```
Otherwise, if I switched `ThreadLocalScopeManager._tls_scope` to be the `_context` property that returns tls, `FlaskScopeManager` would only need to implement that.

**Comment by rmfitzpatrick** [comment]
Thanks for the review, @carlosalberto, but I'm not sure I understand your testbed example suggestion.  Are you requesting that I remove the integration tests?

File: opentracing/scope_managers/flask.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+from .constants import ACTIVE_ATTR
+
+
+class FlaskScopeManager(ScopeManager):
+    """
+    :class:`~opentracing.ScopeManager` implementation for **Flask**
+    that stores the :class:`~opentracing.Scope` in a Flask
+    :class:`RequestContext`, made accessible via
+    :attr:`flask._request_ctx_stack`.
+    """
```
Comment (yurishkuro) [inline opentracing/scope_managers/flask.py]: - Why is this needed?  
- When would I want to use that instead of normal thread-local ScopeManager? What benefits does this one provide?
- We have a flask project under contrib, what's the reason not to have this defined there?

Comment (rmfitzpatrick) [reply opentracing/scope_managers/flask.py]: @yurishkuro, though I can't speak from direct experience of tls not being adequate, according to the [Werkzeug docs](http://werkzeug.pocoo.org/docs/0.14/local/) thread-local storage isn't reliable in all use cases/configurations and using the provided mechanism that guarantees request-local storage seems sound for instrumented Flask apps:
```
[...] Also, whether every request gets its own thread is not guaranteed in WSGI. It could be that a request is reusing a thread from a previous request, and hence data is left over in the thread local object.
```

As to the location of the scope manager, I was keeping in line with what appears to be the current practice of including them in this repo (like Tornado's).  I'd be happy to move it to contrib's project.

Comment (rmfitzpatrick) [reply opentracing/scope_managers/flask.py]: I'm not sure if the gevent scope manager is also subject to similar issues, but the werkzeug context locals seem to provide request-local storage for it as well.

**Comment by carlosalberto** [comment]
@rmfitzpatrick Hey hey.

Was thinking that @yurishkuro is right, as probable putting this custom, well tuned Flask scope manager could go into https://github.com/opentracing-contrib/python-flask (for the `v1.0.0` branch, or to `master` after those changes are merged)

The reason of this, is that the general case for Flask is covered by `ThreadLocalScopeManager`, and having this implementation is an optimization we can offer at the instrumentation level.

Also, in that case, you could leave the stuff under `testbed` out ;) 

Would you mind creating a PR against the flask instrumentation repo?

**Comment by rmfitzpatrick** [comment]
Moved to https://github.com/opentracing-contrib/python-flask/pull/26.  Thanks for the consideration!
